src/shogun/kernel/SpectralMixtureKernel.cpp

```diff
@@ -0,0 +1,221 @@
+
+	return result;
+}
+
+SGVector<float64_t> CSpectralMixtureKernel::get_squared_features(CDotFeatures* features)
+{
+	SGVector<float64_t> sqr_features = SGVector<float64_t>(features->get_num_vectors());
+	
+	for (int i=0; i<sqr_features.size(); i++)
+		sqr_features[i] = features->dot(i, features, i);
+
+	return sqr_features;
+}
+
+SGMatrix<float64_t> CSpectralMixtureKernel::get_square_distance_matrix()
+{
+	CDotFeatures* dot_lhs = dynamic_cast<CDotFeatures*>(lhs);
+	CDotFeatures* dot_rhs = dynamic_cast<CDotFeatures*>(rhs);
+
+	REQUIRE(dot_lhs != NULL, "Left-hand-side features must be of type CDotFeatures");
+	REQUIRE(dot_rhs != NULL, "Right-hand-side features must be of type CDotFeatures");
+	
+	SGMatrix<float64_t> lhs_mtrx = dot_lhs->get_computed_dot_feature_matrix();
+	SGMatrix<float64_t> rhs_mtrx = dot_rhs->get_computed_dot_feature_matrix();
+
+	/* rhs^T . lhs: index (i, j) is dot product of lhs_i and rhs_j.
+	   result scaled by minus two for distance computation */
+	SGMatrix<float64_t> dot_mtrx = SGMatrix<float64_t>::matrix_multiply(rhs_mtrx, lhs_mtrx, true, false, -2.0);
+
+	SGVector<float64_t> sqr_lhs = get_squared_features(dot_lhs);
+	SGVector<float64_t> sqr_rhs = get_squared_features(dot_rhs);
+
+	for (int i=0; i<dot_mtrx.num_cols; i++)
+	{
+		/* To each column vector is added a scalar: lhs_i^2
+		   and the vector [rhs_1^2 ... rhs_n^2]^T*/
+		float64_t* col_vector_ptr = dot_mtrx.get_column_vector(i);
+
+		SGVector<float64_t>::add_scalar(sqr_lhs[i], col_vector_ptr, dot_mtrx.num_rows);
+		SGVector<float64_t>::add(col_vector_ptr, 1.0, col_vector_ptr, 1.0, sqr_rhs, dot_mtrx.num_rows);
+	}
+	
+	return dot_mtrx;
+}
+
+void CSpectralMixtureKernel::compute_weight_derivative(SGMatrix<float64_t>& squared_distances, index_t index)
+{
+	for (int i=0; i<squared_distances.num_rows; i++)
+	{
+		for (int j=0; j<squared_distances.num_cols; j++)
+		{
+			// 2*w_i*exp(-2*pi^2*dist^2*sigma_i^2) * cos(2*pi*dist*mu_i)
+			float64_t sqr_dist = squared_distances(i,j);
+			float64_t exp_factor = CMath::exp(-2*CMath::pow(CMath::PI,2)*sqr_dist*CMath::pow(m_stddeviations[index],2));
+			float64_t cos_factor = CMath::cos(2*CMath::PI*CMath::sqrt(sqr_dist)*m_means[index]);
+			squared_distances(i, j) = 2*m_weights[index]*exp_factor*cos_factor;
+		}
+	}
+}
+
+void CSpectralMixtureKernel::compute_mean_derivative(SGMatrix<float64_t>& squared_distances, index_t index)
+{
+	for (int i=0; i<squared_distances.num_rows; i++)
+	{
+		for (int j=0; j<squared_distances.num_cols; j++)
+		{
+			// -2*pi*dist*mu_i^2*exp(-2*pi^2*dist^2*sigma_i^2)*sin(2*pi*dist*mu_i)
+			float64_t sqr_dist = squared_distances(i,j);
+			float64_t outer_factor = -2*CMath::PI*CMath::sqrt(sqr_dist)*CMath::pow(m_weights[index],2);
+			float64_t exp_factor = CMath::exp(-2*CMath::pow(CMath::PI, 2)*sqr_dist*CMath::pow(m_stddeviations[index], 2));
```

Comment:
@karlnapf shouldn't be much work. I think openmp might turn out to be an overkill for native since modern compiler generates vectororized instructions for the loops. I have been using `std::transform` and `std::for_all` with lambda which I believe are reasonably fast. Maybe a benchmark will help.

I am worried about implicit conversion to/from SG/CGPU structs. They are performance killers and they are obstacles for in-place operations as well. Not sure if we should keep those (or maybe handle in a different way). Will think about it a bit.